File: src/ts/src/controlCommand.ts
import {AirplaneCore} from './airplaneCore';
import {sendCmd, sendCmdVolatile} from './httpLayer';

export class AirplaneController extends AirplaneCore {
    count: number = 1;
    private _sendCmdFn: (s: string) => Promise<any> = sendCmd;

    /**
     * 设置命令是否使用非阻塞模式
     *
     * 在 TS 中，我们通常使用 Promise。fastMode 决定使用 volatile 命令，
     * futureMode 决定是否立即返回 Promise (在 TS 中始终返回 Promise，但可以通过不 await 来实现非阻塞)
     */
    useFastMode(fastMode: boolean = true, futureMode: boolean = true) {
        if (fastMode) {
            this._sendCmdFn = sendCmdVolatile;
        } else {
            this._sendCmdFn = sendCmd;
        }
    }

    private _nextCount(): number {
        this.count = this.count + 2;
        return this.count;
    }

    private _prepareCommand(command: string): string {
        return `${this.keyName} ${this._nextCount()} ${command}`;
    }

    protected async _sendCmd(command: string): Promise<any> {
        return this._sendCmdFn(this._prepareCommand(command));
    }

    mode(m: number) {
        return this.airplaneMode(m);
    }

    takeoff(high: number) {
        return this._sendCmd(`takeoff ${high}`);
    }

    land() {
        return this._sendCmd(`land`);
    }

    emergency() {
        return this._sendCmd(`emergency`);
    }

    up(distance: number) {
        return this._sendCmd(`up ${distance}`);
    }

    down(distance: number) {
        return this._sendCmd(`down ${distance}`);
    }

    forward(distance: number) {
        return this._sendCmd(`forward ${distance}`);
    }

    back(distance: number) {
        return this._sendCmd(`back ${distance}`);
    }

    left(distance: number) {
        return this._sendCmd(`left ${distance}`);
    }

    right(distance: number) {
        return this._sendCmd(`right ${distance}`);
    }

    goto(x: number, y: number, h: number) {
        return this._sendCmd(`goto ${x} ${y} ${h}`);
    }

    flip(direction: string) {
        return this._sendCmd(`flip ${direction} 1`);
    }

    flipForward() {
        return this.flip("f");
    }

    flipBack() {
        return this.flip("b");
    }

    flipLeft() {
        return this.flip("l");
    }

    flipRight() {
        return this.flip("r");
    }

    rotate(degree: number) {
        return this._sendCmd(`rotate ${degree}`);
    }

    cw(degree: number) {
        return this._sendCmd(`cw ${degree}`);
    }

    ccw(degree: number) {
        return this._sendCmd(`ccw ${degree}`);
    }

    high(h: number) {
        return this._sendCmd(`high ${h}`);
    }

    speed(s: number) {
        return this._sendCmd(`setSpeed ${s}`);
    }

    led(r: number, g: number, b: number) {
        return this._sendCmd(`light ${r} ${g} ${b}`);
    }

    bln(r: number, g: number, b: number) {
        return this._sendCmd(`bln ${r} ${g} ${b}`);
    }

    rainbow(r: number, g: number, b: number) {
        return this._sendCmd(`rainbow ${r} ${g} ${b}`);
    }

    airplaneMode(m: number) {
        return this._sendCmd(`airplane_mode ${m}`);
    }

    stop() {
        return this.hover();
    }

    hover() {
        return this._sendCmd(`hover`);
    }
}

File: src/lua/control_command.lua
local airplane_core = require("airplane_core")
local http_layer = require("http_layer")

local AirplaneController = {}
AirplaneController.__index = AirplaneController
setmetatable(AirplaneController, {__index = airplane_core.AirplaneCore})

function AirplaneController.new(data)
    local self = airplane_core.AirplaneCore.new(data)
    setmetatable(self, AirplaneController)
    self.count = 1
    self._send_cmd_fn = http_layer.send_cmd
    return self
end

function AirplaneController:use_fast_mode(fast_mode, future_mode)
    if fast_mode == nil then fast_mode = true end
    if future_mode == nil then future_mode = true end
    
    if not future_mode then
        if fast_mode then
            self._send_cmd_fn = http_layer.send_cmd_volatile
        else
            self._send_cmd_fn = http_layer.send_cmd
        end
    else
        -- Lua doesn't have a built-in ThreadPoolExecutor like Python.
        -- In a real Lua environment, this might use a coroutine or a task scheduler.
        -- For this preliminary implementation, we'll just wrap it.
        if fast_mode then
            self._send_cmd_fn = function(cmd)
                -- Placeholder for async/future execution
                return http_layer.send_cmd_volatile(cmd)
            end
        else
            self._send_cmd_fn = function(cmd)
                -- Placeholder for async/future execution
                return http_layer.send_cmd(cmd)
            end
        end
    end
end

function AirplaneController:_next_count()
    self.count = self.count + 2
    return self.count
end

function AirplaneController:_prepare_command(command)
    return self.keyName .. " " .. tostring(self:_next_count()) .. " " .. command
end

function AirplaneController:_send_cmd(command)
    return self._send_cmd_fn(self:_prepare_command(command))
end

function AirplaneController:mode(mode_val)
    self:airplane_mode(mode_val)
end

function AirplaneController:takeoff(high)
    return self:_send_cmd(string.format("takeoff %d", high))
end

function AirplaneController:land()
    return self:_send_cmd("land")
end

function AirplaneController:emergency()
    return self:_send_cmd("emergency")
end

function AirplaneController:up(distance)
    return self:_send_cmd(string.format("up %d", distance))
end

function AirplaneController:down(distance)
    return self:_send_cmd(string.format("down %d", distance))
end

function AirplaneController:forward(distance)
    return self:_send_cmd(string.format("forward %d", distance))
end

function AirplaneController:back(distance)
    return self:_send_cmd(string.format("back %d", distance))
end

function AirplaneController:left(distance)
    return self:_send_cmd(string.format("left %d", distance))
end

function AirplaneController:right(distance)
    return self:_send_cmd(string.format("right %d", distance))
end

function AirplaneController:goto_pos(x, y, h)
    return self:_send_cmd(string.format("goto %d %d %d", x, y, h))
end

function AirplaneController:flip(direction)
    return self:_send_cmd(string.format("flip %s 1", direction))
end

function AirplaneController:flip_forward()
    self:flip("f")
end

function AirplaneController:flip_back()
    self:flip("b")
end

function AirplaneController:flip_left()
    self:flip("l")
end

function AirplaneController:flip_right()
    self:flip("r")
end

function AirplaneController:rotate(degree)
    return self:_send_cmd(string.format("rotate %d", degree))
end

function AirplaneController:cw(degree)
    return self:_send_cmd(string.format("cw %d", degree))
end

function AirplaneController:ccw(degree)
    return self:_send_cmd(string.format("ccw %d", degree))
end

function AirplaneController:high(high_val)
    return self:_send_cmd(string.format("high %d", high_val))
end

function AirplaneController:speed(speed_val)
    return self:_send_cmd(string.format("setSpeed %d", speed_val))
end

function AirplaneController:led(r, g, b)
    return self:_send_cmd(string.format("light %d %d %d", r, g, b))
end

function AirplaneController:bln(r, g, b)
    return self:_send_cmd(string.format("bln %d %d %d", r, g, b))
end

function AirplaneController:rainbow(r, g, b)
    return self:_send_cmd(string.format("rainbow %d %d %d", r, g, b))
end

function AirplaneController:airplane_mode(mode_val)
    return self:_send_cmd(string.format("airplane_mode %d", mode_val))
end

function AirplaneController:stop()
    return self:hover()
end

function AirplaneController:hover()
    return self:_send_cmd("hover")
end

return {
    AirplaneController = AirplaneController
}
